### PR TITLE
Smh0109/issue17

### LINKED ISSUE #17: the point of intersection of two hyperbolic line segments
Hello, I would be grateful for any contribution to the discussion on this issue.

I want to determine the intersection of two hyperbolic line-segments, e.g. _AB_ and _CD_ given by these endpoints. But the command _AB.intersectionsWithHcycle.CD_ takes _AB_ and _CD_ as hyperbolic lines. So the result is intersection of the hyperbolic lines _AB_, _CD_ (if it exists) even if the hyperbolic line-segments do not intersect. Is there any way how to fix that? 

Of course, analogous to Euclidean geometry, we could then verify whether the resulting point is the inner point of hyperbolic line-segment _AB_ and _CD_. Let us say, that hyperbolic lines _AB_, _CD_ do intersect in point _X_. Maybe using this command: _R_=hyper.poincare.Point.Point.midpointWith(_A_, _B_, frac = _k_), we can look for such a value of _k_, that _X_ = _R_. But my experiences with Python are very weak. The problem goes further, if e.g. point _D_ is an ideal point, so _CD_ is a hyperbolic ray.

### PR DESCRIPTION
**Pull request by smh0109**
I guess those are the implementations needed.
Do you have anything I schould do differently?
I must admit, this is my first contribution. As an undergard my coding experience is very limited.

Closes #17.

### CODE REVIEW

**Comment by cduck** [comment]
Thanks.  This is a great first PR.  I'll try it out and get back to you with comments soon.  (I added "Closes #17" so GitHub will automatically close that issue once this PR is merged.)

File: hyperbolic/euclid/Arc.py
Comment (cduck) [inline hyperbolic/euclid/Arc.py:53]: I tried this out and it incorrectly returns False
- when `startDeg > endDeg` and cw=False
- when `startDeg < endDeg` and cw=True
- when the point is on the lower half of the circle (atan2 returns a negative number but startDeg and endDeg are normalized to positive numbers).

Try changing `start`, `end`, and `cw` in this code to test it out:
```python
import drawSvg as draw
from hyperbolic import euclid
import math

start = 75
end = 230
cw = False

x, y, r = 0.2, 0.3, 0.4
circle = euclid.shapes.Circle(x, y, r)
arc = euclid.shapes.Arc(x, y, r, start, end, cw=cw)
print('Drawn shape:', arc)

points = [
    arc.startPoint(),
    arc.endPoint(),
] + [
    (x + r*math.cos(math.radians(deg)),
     y + r*math.sin(math.radians(deg)))
    for deg in range(0, 360, 20)
]

d = draw.Drawing(2.1, 2.1, origin='center')

d.draw(circle, fill='none', stroke_width=0.02, stroke='green')
d.draw(arc, fill='black')
for x, y in points:
    # Blue only if on segment
    fill = 'skyblue' if arc.isPointOnSegment(x, y) else 'red'
    p = euclid.shapes.Circle(x, y, 0.02)
    d.draw(p, fill=fill)

d.setRenderSize(w=400)
d
```

File: hyperbolic/euclid/Arc.py
Comment (cduck) [inline hyperbolic/euclid/Arc.py:44]: When adding code to an existing project, it is best to try to match the same code style.  I started this project a while ago so the style doesn't conform to [PEP8](https://pep8.org/) unfortunately.
```suggestion
        ''' Assumes that Point is on the circle '''
```

Thanks for adding docstrings.  Consider making them more descriptive for users who haven't looked at the code.
```suggestion
        ''' Assumes that the given point is on the circle and returns True if on
        this arc '''
```

File: hyperbolic/euclid/Circle.py
Comment (cduck) [inline hyperbolic/euclid/Circle.py:32]: Another suggestion:
```suggestion
        ''' Assumes that the given point is on the circle and always returns
        True '''
```

File: hyperbolic/euclid/Line.py
Comment (cduck) [inline hyperbolic/euclid/Line.py:70]: When you have long comments, put them on the line before.  Also, did you mean `startpoint->endPoint` not `startpoint->endPoint->(x,y)`?
```suggestion
        # Dot product of vector startPoint->(x,y) and startpoint->endPoint
        K1=(x-self.x1)*(self.x2-self.x1)+(y-self.y1)*(self.y2-self.y1)
        # Dot product of vector endPoint->(x,y) and endpoint->startPoint
        K2=(x-self.x2)*(self.x1-self.x2)+(y-self.y2)*(self.y1-self.y2)
```

File: hyperbolic/euclid/Line.py
Comment (cduck) [inline hyperbolic/euclid/Line.py:74]: Simpler suggestion:
```suggestion
        return K1>=0 and K2>=0
```

File: hyperbolic/euclid/Line.py
Comment (cduck) [inline hyperbolic/euclid/Line.py:69]: Nice.

File: hyperbolic/poincare/Hypercycle.py
Comment (cduck) [inline hyperbolic/poincare/Hypercycle.py:94]: This is a bug.  `pop` returns the item removed, not the new list.  The list is modified so you don't need to reassign it.

My test code:
```python
import drawSvg as draw
from drawSvg import Drawing
from hyperbolic import euclid
from hyperbolic.poincare.shapes import *

p1 = Point(-.2,-.5)
p2 = Point(.2, .5)
line = Line.fromPoints(*p1, *p2, segment=False)
segment = Line.fromPoints(*p1, *p2, segment=True)
print('Drawn shape:', segment.projShape)

segment2 = Line.fromPoints(-.7, 0, .5, .6, segment=True)
#segment2 = Line.fromPoints(-.7, 0, .2, .6, segment=True)  # <- Triggers bug
print('Drawn shape:', segment2.projShape)

points = [
    segment.startPoint().midpointWith(segment.endPoint(), frac=x)
    for x in [-0.2, 0, 0.2, 0.5, 0.9, 1, 1.3]
]

d = Drawing(2.1, 2.1, origin='center')
d.draw(euclid.shapes.Circle(0, 0, 1), fill='#eee')

d.draw(segment, hwidth=0.15, fill='black')
d.draw(segment2, hwidth=0.15, fill='gray')
for p in points:
    # Blue only if on segment
    fill = 'skyblue' if segment.isPointOnSegment(*p) else 'red'
    d.draw(p, hradius=0.05, fill=fill)
intersects = segment.segmentIntersectionsWithHcycle(segment2)
for p in intersects:
    print('Intersects!')
    d.draw(p, hradius=0.05, fill='white')

d.setRenderSize(w=400)
d
```

**Review by cduck — CHANGES_REQUESTED** [review]
This is a good start.  In the comments below, I point out two issues with the code and several suggestions for better code style.

When you are done making changes, notify me by replying below and I will review again.  It's normal for this review process to repeat several times before I merge it.

File: hyperbolic/euclid/Arc.py
Comment (smh0109) [reply hyperbolic/euclid/Arc.py]: I changed the code, and tested it, seems to work in all cases now.
In your code you used isPointOnSegment(self, x, y, tol) is a tolerance a feature needed/wanted? Or is it for testing only?

Comment (cduck) [reply hyperbolic/euclid/Arc.py]: I forgot to mention tolerance.  I noticed some cases where either the start or endpoint was not on the segment because of rounding error.  This would be a good feature (with a default tolerance of something small) but I'd do that in a follow-up PR.

**Comment by smh0109** [comment]
Does it take you longer to review the code or did I miss something important?
If you were just waiting for a comment as a review request, this is it.

**Comment by cduck** [comment]
You are good.  I've just been too busy.  I will try to find time this week.

File: hyperbolic/euclid/Arc.py
Comment (cduck) [inline hyperbolic/euclid/Arc.py:49]: Don't need the else clause.  It doesn't do anything.
```suggestion
```

**Review by cduck — APPROVED** [review]
Looks great!  There is just one minor thing to fix and I'll merge it.  Thanks for your patience.

**Comment by smh0109** [comment]
Awsome!
I'm thankful I've been able to contribute.
As I'm doing a project about delta-slim triangles I created a subclass Triangle(Polygon) which I'll post in the next few days as a Isssue...

**Comment by cduck** [comment]
Published [1.4.0](https://pypi.org/project/hyperbolic/1.4.0/).

**Pull request merged**
